x/oracle/keeper/grpc_query.go

```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
+	reporterPubKey, err := sdk.GetPubKeyFromBech32(sdk.Bech32PubKeyTypeAccPub, req.Reporter)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("unable to get reporter's public key: %s", err.Error()))
+	}
+	requestVerificationContent := types.NewRequestVerification(req.ChainId, validator, types.RequestID(req.RequestId), types.ExternalID(req.ExternalId))
+	signByte := requestVerificationContent.GetSignBytes()
+	if !reporterPubKey.VerifySignature(signByte, req.Signature) {
+		return nil, status.Error(codes.InvalidArgument, "invalid reporter's signature")
+	}
+
+	// Provided reporter should be authorized by the provided validator
+	reporters := k.GetReporters(ctx, validator)
+	reporter := sdk.AccAddress(reporterPubKey.Address().Bytes())
+	isReporterAuthorizedByValidator := false
+	for _, existingReporter := range reporters {
+		if reporter.Equals(existingReporter) {
+			isReporterAuthorizedByValidator = true
+			break
+		}
+	}
+	if !isReporterAuthorizedByValidator {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("%s is not an authorized reporter of %s", reporter, req.Validator))
+	}
+
+	// Provided request should exist on chain
+	request, err := k.GetRequest(ctx, types.RequestID(req.RequestId))
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("unable to get request from chain: %s", err.Error()))
+	}
+
+	// Provided validator should be assigned to response to the request
+	isValidatorAssigned := false
+	for _, requestedValidator := range request.RequestedValidators {
+		v, _ := sdk.ValAddressFromBech32(requestedValidator)
+		if validator.Equals(v) {
+			isValidatorAssigned = true
+			break
+		}
+	}
+	if !isValidatorAssigned {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("%s is not assigned for request ID %d", validator, req.RequestId))
+	}
+
+	// Provided external ID should be required by the request determined by oracle script
+	var dataSourceID *types.DataSourceID
+	for _, rawRequest := range request.RawRequests {
+		if rawRequest.ExternalID == types.ExternalID(req.ExternalId) {
+			dataSourceID = &rawRequest.DataSourceID
+			break
+		}
+	}
+	if dataSourceID == nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("no data source required by the request %d found which relates to the external data source with ID %d.", req.RequestId, req.ExternalId))
+	}
+
+	// Provided validator should not have reported data for the request
+	reports := k.GetReports(ctx, types.RequestID(req.RequestId))
+	isValidatorReported := false
+	for _, report := range reports {
+		reportVal, _ := sdk.ValAddressFromBech32(report.Validator)
+		if reportVal.Equals(validator) {
+			isValidatorReported = true
+			break
+		}
+	}
+	if isValidatorReported {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("validator %s already submitted data report for this request", validator))
```

Comment:
Let's think which error code suit this case.

Reply:
Can it be `AlreadyExists`?